Zotero/Scenes/Detail/Lookup/ViewModels/LookupActionHandler.swift

```diff
@@ -11,90 +11,94 @@ import Foundation
-        self.fileStorage = fileStorage
-        self.dbStorage = dbStorage
-        self.translatorsController = translatorsController
-        self.schemaController = schemaController
-        self.dateParser = dateParser
-        self.remoteFileDownloader = remoteFileDownloader
+    init(identifierLookupController: IdentifierLookupController) {
+        self.identifierLookupController = identifierLookupController
         self.disposeBag = DisposeBag()
     }
 
     func process(action: LookupAction, in viewModel: ViewModel<LookupActionHandler>) {
+        let collectionKeys = viewModel.state.collectionKeys
+        let libraryId = viewModel.state.libraryId
         switch action {
-        case .initialize(let webView):
-            let handler = LookupWebViewHandler(webView: webView, translatorsController: self.translatorsController)
-            self.lookupWebViewHandler = handler
-
-            handler.observable
-                   .observe(on: MainScheduler.instance)
-                   .subscribe(with: viewModel, onNext: { [weak self] viewModel, result in
-                       self?.process(result: result, in: viewModel)
-                   })
-                   .disposed(by: self.disposeBag)
-
-            self.remoteFileDownloader.observable
-                .observe(on: MainScheduler.instance)
-                .subscribe(with: viewModel, onNext: { [weak self] viewModel, update in
-                    switch update.kind {
-                    case .ready(let attachment):
-                        self?.finish(download: update.download, attachment: attachment, in: viewModel)
-                    case .cancelled, .failed, .progress: break
+        case .initialize:
+            identifierLookupController.initialize(libraryId: libraryId, collectionKeys: collectionKeys) { [weak self] lookupData in
+                guard let self, let lookupData else {
+                    DDLogError("LookupActionHandler: can't create observer")
+                    return
+                }
+                if viewModel.state.restoreLookupState, !lookupData.isEmpty {
+                    DDLogInfo("LookupActionHandler: restoring lookup state")
+                    self.update(viewModel: viewModel) { state in
+                        state.lookupState = .lookup(lookupData)
+                    }
+                }
+                self.identifierLookupController.observable
+                    .observe(on: MainScheduler.instance)
+                    .subscribe(with: viewModel) { [weak self] viewModel, update in
+                        guard let self else { return }
+                        switch update.kind {
+                        case .lookupError(let error):
+                            self.update(viewModel: viewModel) { state in
+                                state.lookupState = .failed(error)
+                            }
+
+                        case .noIdentifiersDetected:
```

Comment:
`.lookupError` is used for errors observed from `LookupWebViewHandler`, while `.noIdentifiersDetected` is not considered an error in the `IdentifierLookupController` context. I modified it in `LookupActionHandler`, because it is handled as an error, to display the failure message. Do you think we should merge them both in one case?

Reply:
I'd maybe consider no identifiers detected as error, because the input was most likely wrong then. But if there are valid reasons why it's not detected while not being an error this is fine then.

Reply:
Check https://github.com/zotero/zotero-ios/pull/707/commits/a29a864c86ff0eabc79981d05550e39492fdba52

It merges the `identifiersDetected` and `noIdentifiersDetected` states, and lets `LookupActionHandler` to determine how the view model should be affected when the identifiers array passed with the update is empty.